### PR TITLE
Code generation for u32 arithmetic

### PR DESCRIPTION
**Pull request by jjcnn**
The code generation pass is now ready for review.

One thing that is definitely missing is connecting the pass to the executable, but I will need some help with that.

I also need to make a couple of changes in the miden-vm repo, since the assembly instructions we are creating are not exported by MidenVM.

### CODE REVIEW

File: ir/src/codegen/codegen.rs
Comment (bitwalker) [inline ir/src/codegen/codegen.rs:411]: I would pass the entire `DataFlowGraph` in here, as it is the only way to get instruction results, which it appears are currently not being handled.

I don't think this will generate working code at the moment - the reason being that while the result of an expression is on the Miden operand stack, that value is not being stored into the local corresponding to the instruction result(s) in Miden IR. Thus, the next instruction that tries to use that local will get garbage (likely `0` in simple programs, as the memory hasn't been allocated into yet) instead of the expected value.

For example, consider the following function in Miden IR:

```
pub fn multiply_and_add(i32, i32, i32): i32 {
entry(a: i32, b: i32, c: i32):
    v0 = mul a, b
    v1 = add v0, c
    ret v1
}
```

The current lowering does not store the result of `mul a, b` in `v0`, so the subsequent `add v0, c` is non-deterministic/incorrect. Similarly, the `add` doesn't store its result in `v1`, so the `ret` will just return whatever value is in the memory allocated for `v1`.

**Review by bitwalker — CHANGES_REQUESTED** [review]
Nice @jjcnn! You've gotten up to speed with the IR and Rust quite quickly 😃. I've left a handful of comments, mostly minor things, with only one major issue that will need to be addressed.

I think you've got the gist of things worked out here, but there are some key things missing that will come up eventually. In particular, one of the things we want to be able to do is reduce the amount of operations related to loading/storing memory for locals as much as possible, or realistic compiled programs will have tons of unnecessary instructions, resulting in really large proofs. 

To handle this, we need to track the operand stack during codegen, and track where instruction operands and results are stored from instruction to instruction. For example, if we have a sequence of arithmetic instructions, it is guaranteed that the result at each step will be in the "accumulator" register (which in MASM is the top of the operand stack). This means that we only need to potentially load/store one of the operands, or better yet, use one of the stack manipulation instructions to move operands on the operand stack into their expected locations. This approach is referred to in the literature as "destination-driven" codegen, and is especially well suited to generating assembly. There are also extensions which include control context (i.e. where to transfer control after an instruction), which is likely also useful for us, but the most important bits have to do with the data/operands. I have some links handy if you'd like to dig into how that approach to codegen works in detail.

In any case this PR represents a solid initial sketch, just needs a bit more fleshing out and we'll be in business!

**Comment by jjcnn** [comment]
This PR deliberately does not attempt to optimize load and store operations, because it requires a liveness analysis to do so. To see why, consider the following program:

```
y = a + b;
x1 = y + z;
x2 = y + w;
```
This situation will occur whenever a local variable from the source program is read multiple times.

If we naïvely store the value of `y` on the stack, then the value will be consumed when calculating `x1`, and will not be available when calculating `x2`. If we want to use the stack to store `y`, we therefore need to either
- Know that `y` i still live, so that we can perform a `dup` before we calculate `x1`, or
- Assume that `y` may still be live, so that we can perform a `dup` before we calculate `x1`, followed by a `pop` if `y` turns out to not be live, or
- Assume that `y` may be live, perform a `store` operation, and then perform a `load` operation for each use of `y`, as is done in this PR.

Optimizing load and store operations away definitely needs to happen, but it is also definitely scope creep if we do it in this PR. What we really need at the moment is an executable compiler that generates executable code. Optimizing the produced code can happen later.

**Comment by bitwalker** [comment]
@jjcnn I'm not suggesting we should optimize at this point, but rather that the code generator itself will need to be structured a particular way in order to be able to do it at all, and that doesn't require liveness analysis. It is absolutely fine to assume that all values are live until the end of the function at this stage, and introduce stack ops or loads/stores to keep values live, but I would like to make sure that we at least have the general structure set up in a generally destination-driven style.

I'm fine with that happening in a follow-on PR, doesn't need to be done here per se, but I'd suggest making that the next step. It will be far easier to make a change like that while the implementation is in its early stages, rather than once you've put more work.

As far as putting all of the pieces together in the driver, that's the easy part, so if you'd like to have that for your work on this, I can get that set up as soon as this is merged.

**Comment by jjcnn** [comment]
> @jjcnn I'm not suggesting we should optimize at this point, but rather that the code generator itself will need to be structured a particular way in order to be able to do it at all, and that doesn't require liveness analysis. It is absolutely fine to assume that all values are live until the end of the function at this stage, and introduce stack ops or loads/stores to keep values live, but I would like to make sure that we at least have the general structure set up in a generally destination-driven style.
> 
> I'm fine with that happening in a follow-on PR, doesn't need to be done here per se, but I'd suggest making that the next step. It will be far easier to make a change like that while the implementation is in its early stages, rather than once you've put more work.
> 

I think you sent me some links on this already, but I haven't had a chance to look at them yet. The rough idea I have in my head of how to do it is as follows:

- Do a separate abstract-interpretation pass after (or as part of) the liveness analysis, but before code generation. The interpretation models the stack as if all values are stack values rather than locals. If a value is ever read when it is not in the top 16 elements on the stack (because we only have access to the top 16 elements), then we mark it as a local variable. (This pass can also counts the number of needed locals, so that we can generate a tighter bound on the number of locals in the procedure signature).
- During code generation:
- - When we produce a value
- - - If it is marked as a local, then store it.
- - - If it not makred as a local, then leave it on the stack
- - When we read a value
- - - If it is marked as a local, then load it.
- - - If it is not marked as a local, then dup or swap it from its current position on the stack (dup or swap depending on whether it's live after the current instruction.

The only structural change we would need to make to the code generation pass is that we would need to pass the current stack of values as an argument when calling `codegen_instruction`.

**Comment by bitwalker** [comment]
> I think you sent me some links on this already, but I haven't had a chance to look at them yet. 

Yeah I'd definitely give those a look, particularly the one that I mentioned has an easier-to-digest description of the approach, I think that will better illustrate how to approach things in our context. Something to note about that link is that its example code is targeting x86 assembly, which of course is a register machine - however the general mechanism works basically the same for any register or stack machine, the primary difference is in how you track what the possible destinations are. For us, we'd probably want to treat the first 16 slots of the stack somewhat like registers, in that they can be addressed as destinations directly, similar to a register, and then everything else gets loaded/stored to linear memory.

> The rough idea I have in my head of how to do it is as follows: ...

Destination-driven codegen is indeed a form of abstract interpretation, except it specifically models where inputs/outputs are stored at every instruction. More precisely, when generating each instruction, the handler for that instruction gets metadata that tells it where each of its arguments are located (i.e. what "destination" they were stored to), as well as what destination(s) to use for storing the results of the instruction. With the aid of liveness analysis, it is possible to reuse destinations for inputs/outputs (as is a common pattern with the "accumulator" register on typical register-machine archcitectures). It is also possible to have a destination of "nowhere", i.e. the value can be discarded because it is never used. 

By using an auxiliary structure that maps SSA values to destinations, you can get away without liveness analysis, because you allocate destinations as you go along, so if you, for example, use up all 16 slots of the operand stack, then you can "spill" some values to linear memory temporarily, freeing up needed slots until those values are needed again. Each instruction handler can handle spilling/loading as needed at the current point in the program - but in the general case, many sequences of instructions use a very limited set of virtual registers, so spills/loads end up being compartively rare. Of course liveness analysis has a variety of benefits, not least of which is the ability to allocate a minimal set of locals, but it also tells us what values are live at every instruction, which allows us to make much better decisions about how where to store values across instructions - but we can implement a pretty decent code generator without it using this approach.

File: ir/Cargo.toml
Comment (jjcnn) [reply ir/Cargo.toml]: Removed.

File: ir/src/codegen/codegen.rs
Comment (jjcnn) [reply ir/src/codegen/codegen.rs]: Makes sense. I'm still getting used to the `::` notation, so it feels more natural to use renaming. But I'll get the hang of `::` eventually.

File: ir/src/codegen/codegen.rs
Comment (jjcnn) [reply ir/src/codegen/codegen.rs]: Also fixed.

File: ir/src/codegen/codegen.rs
Comment (jjcnn) [reply ir/src/codegen/codegen.rs]: `store` operation added. As you said, the dataflow graph now needs to be passed to `codegen_instruction`.

**Review by bitwalker — APPROVED** [review]

**Comment by jjcnn** [comment]
> @jjcnn Looks like you'll have to rebase locally in order to get this merged, there's conflicts of some sort

Done. Waiting for the CI checks.

**Comment by bitwalker** [comment]
I'm going to close this now that #27 is about to be merged, since that supersedes this PR

**Pull request closed without merging**
